File: src_app/lib/services/image_suggestion_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../config/api_config.dart';

/// Service gợi ý hình ảnh cho flashcard
class ImageSuggestionService {
  /// Lấy danh sách ảnh gợi ý cho từ vựng
  /// [word] - Từ vựng cần tìm ảnh
  /// [count] - Số lượng ảnh (mặc định 5, tối đa 10)
  static Future<ImageSuggestionResult> suggestImages(String word, {int count = 5}) async {
    try {
      final uri = Uri.parse('${ApiConfig.baseUrl}/api/images/suggest')
          .replace(queryParameters: {
        'word': word,
        'count': count.toString(),
      });

      print('🖼️ Getting $count image suggestions for: $word');

      final response = await http.get(uri, headers: {
        'Content-Type': 'application/json',
        'ngrok-skip-browser-warning': 'true',
      });

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return ImageSuggestionResult.fromJson(data);
      } else {
        throw Exception('Failed to get image suggestions: ${response.statusCode}');
      }
    } catch (e) {
      print('❌ Image suggestion error: $e');
      rethrow;
    }
  }

  /// Kiểm tra trạng thái service
  static Future<Map<String, dynamic>> getStatus() async {
    try {
      final uri = Uri.parse('${ApiConfig.baseUrl}/api/images/status');

      final response = await http.get(uri, headers: {
        'ngrok-skip-browser-warning': 'true',
      });

      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        throw Exception('Failed to get status');
      }
    } catch (e) {
      print('❌ Image service status error: $e');
      rethrow;
    }
  }
}

/// Kết quả gợi ý hình ảnh
class ImageSuggestionResult {
  final bool success;
  final String? message;
  final String? word;
  final int totalFound;
  final List<ImageInfo> images;

  ImageSuggestionResult({
    required this.success,
    this.message,
    this.word,
    required this.totalFound,
    required this.images,
  });

  factory ImageSuggestionResult.fromJson(Map<String, dynamic> json) {
    return ImageSuggestionResult(
      success: json['success'] ?? false,
      message: json['message'],
      word: json['word'],
      totalFound: json['totalFound'] ?? 0,
      images: (json['images'] as List<dynamic>?)
          ?.map((e) => ImageInfo.fromJson(e))
          .toList() ??
          [],
    );
  }
}

/// Thông tin ảnh
class ImageInfo {
  final int? id;
  final String url;           // URL mặc định (medium)
  final String? original;     // Full resolution
  final String? large;        // 940px wide
  final String? medium;       // 350px wide
  final String? small;        // 130px wide
  final String? tiny;         // 100x100
  final String? photographer;
  final String? photographerUrl;
  final String? alt;

  ImageInfo({
    this.id,
    required this.url,
    this.original,
    this.large,
    this.medium,
    this.small,
    this.tiny,
    this.photographer,
    this.photographerUrl,
    this.alt,
  });

  factory ImageInfo.fromJson(Map<String, dynamic> json) {
    return ImageInfo(
      id: json['id'],
      url: json['url'] ?? json['medium'] ?? '',
      original: json['original'],
      large: json['large'],
      medium: json['medium'],
      small: json['small'],
      tiny: json['tiny'],
      photographer: json['photographer'],
      photographerUrl: json['photographerUrl'],
      alt: json['alt'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'url': url,
      'original': original,
      'large': large,
      'medium': medium,
      'small': small,
      'tiny': tiny,
      'photographer': photographer,
      'photographerUrl': photographerUrl,
      'alt': alt,
    };
  }
}

File: src_backend/src/main/java/com/tieuluan/backend/service/ImageSuggestionService.java
package com.tieuluan.backend.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.Data;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.*;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

/**
 * Service gợi ý hình ảnh từ Pexels
 * Trả về 5 ảnh để người dùng chọn
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class ImageSuggestionService {

    @Value("${pexels.api.key}")
    private String apiKey;

    @Value("${pexels.api.url:https://api.pexels.com/v1/search}")
    private String apiUrl;

    private final RestTemplate restTemplate = new RestTemplate();
    private final ObjectMapper objectMapper = new ObjectMapper();

    private static final int DEFAULT_IMAGE_COUNT = 5;

    /**
     * Lấy danh sách ảnh gợi ý cho từ vựng
     * @param word Từ vựng cần tìm ảnh
     * @return Danh sách 5 ảnh gợi ý
     */
    public ImageSuggestionResult suggestImages(String word) {
        return suggestImages(word, DEFAULT_IMAGE_COUNT);
    }

    /**
     * Lấy danh sách ảnh gợi ý với số lượng tùy chỉnh
     */
    public ImageSuggestionResult suggestImages(String word, int count) {
        log.info("🖼️ Getting {} image suggestions for: '{}'", count, word);

        ImageSuggestionResult result = new ImageSuggestionResult();
        result.setWord(word);
        result.setImages(new ArrayList<>());

        try {
            if (!isConfigured()) {
                log.error("❌ Pexels API key not configured!");
                result.setSuccess(false);
                result.setMessage("Pexels API chưa được cấu hình");
                return result;
            }

            // Tìm ảnh với từ gốc
            List<ImageInfo> images = searchPexelsImages(word, count);

            // Nếu không đủ ảnh, thử tìm thêm với từ đơn giản hơn
            if (images.size() < count && word.contains(" ")) {
                String simpleWord = word.split(" ")[0];
                log.info("🔄 Not enough images, trying simpler query: '{}'", simpleWord);

                List<ImageInfo> moreImages = searchPexelsImages(simpleWord, count - images.size());
                images.addAll(moreImages);
            }

            // Nếu vẫn không có ảnh, thử với "education" fallback
            if (images.isEmpty()) {
                log.info("🔄 No images found, using generic fallback");
                images = searchPexelsImages("education learning", count);
            }

            result.setImages(images);
            result.setSuccess(!images.isEmpty());
            result.setMessage(images.isEmpty() ? "Không tìm thấy ảnh phù hợp" : "Tìm thấy " + images.size() + " ảnh");
            result.setTotalFound(images.size());

            log.info("✅ Found {} images for '{}'", images.size(), word);
            return result;

        } catch (Exception e) {
            log.error("❌ Error getting image suggestions: {}", e.getMessage(), e);
            result.setSuccess(false);
            result.setMessage("Lỗi khi tìm ảnh: " + e.getMessage());
            return result;
        }
    }

    /**
     * Gọi Pexels API và parse kết quả
     */
    private List<ImageInfo> searchPexelsImages(String query, int count) {
        List<ImageInfo> images = new ArrayList<>();

        try {
            String encodedQuery = URLEncoder.encode(query, StandardCharsets.UTF_8);
            String url = apiUrl + "?query=" + encodedQuery + "&per_page=" + Math.max(count, 5) + "&orientation=landscape";

            log.info("🌐 Calling Pexels: {}", url);

            HttpHeaders headers = new HttpHeaders();
            headers.set("Authorization", apiKey);
            headers.set("User-Agent", "FlashcardApp/1.0");

            HttpEntity<String> request = new HttpEntity<>(headers);
            ResponseEntity<String> response = restTemplate.exchange(url, HttpMethod.GET, request, String.class);

            if (response.getStatusCode() == HttpStatus.OK && response.getBody() != null) {
                JsonNode root = objectMapper.readTree(response.getBody());
                JsonNode photos = root.path("photos");

                if (photos.isArray()) {
                    for (int i = 0; i < Math.min(photos.size(), count); i++) {
                        JsonNode photo = photos.get(i);
                        ImageInfo imageInfo = parsePhotoNode(photo);
                        if (imageInfo != null) {
                            images.add(imageInfo);
                        }
                    }
                }
            }

        } catch (Exception e) {
            log.error("❌ Error searching Pexels: {}", e.getMessage());
        }

        return images;
    }

    /**
     * Parse JSON node thành ImageInfo
     */
    private ImageInfo parsePhotoNode(JsonNode photo) {
        try {
            ImageInfo info = new ImageInfo();

            info.setId(photo.path("id").asLong());
            info.setPhotographer(photo.path("photographer").asText());
            info.setPhotographerUrl(photo.path("photographer_url").asText());
            info.setAlt(photo.path("alt").asText());

            JsonNode src = photo.path("src");
            info.setOriginal(src.path("original").asText());
            info.setLarge(src.path("large").asText());
            info.setMedium(src.path("medium").asText());
            info.setSmall(src.path("small").asText());
            info.setTiny(src.path("tiny").asText());

            // Sử dụng medium làm URL mặc định
            info.setUrl(info.getMedium());

            return info;
        } catch (Exception e) {
            log.error("Error parsing photo: {}", e.getMessage());
            return null;
        }
    }

    /**
     * Kiểm tra API key đã được cấu hình chưa
     */
    public boolean isConfigured() {
        return apiKey != null && !apiKey.isEmpty() && !apiKey.equals("YOUR_API_KEY_HERE");
    }

    // ================== DTOs ==================

    @Data
    public static class ImageSuggestionResult {
        private boolean success;
        private String message;
        private String word;
        private int totalFound;
        private List<ImageInfo> images;
    }

    @Data
    public static class ImageInfo {
        private Long id;
        private String url;           // URL mặc định (medium)
        private String original;      // Full resolution
        private String large;         // 940px wide
        private String medium;        // 350px wide
        private String small;         // 130px wide
        private String tiny;          // 100x100
        private String photographer;
        private String photographerUrl;
        private String alt;           // Mô tả ảnh
    }
}